File: src/types/purchaseBill.ts
export type BillStatus = "draft" | "confirmed" | "paid" | "partial_paid";
export type PurchaseBillPaymentMode = "cash" | "bank_transfer" | "upi" | "cheque" | "credit";
export type PurchaseBillPaymentStatus = "unpaid" | "partial" | "paid";
export type PurchaseBillPaidVia = "cash" | "bank_transfer" | "upi" | "cheque" | "card" | "";
export type PurchaseBillReminderFrequency = "once" | "daily" | "weekly";
export type PurchaseBillReminderChannel = "in_app" | "email" | "both";

export interface PurchaseBillItem {
  id: string;
  productId?: string;
  productName: string;
  sku?: string;
  quantity: number;
  unit: string;
  unitCost: number;
  gstRate: number;
  cgst: number;
  sgst: number;
  igst: number;
  lineTotal: number;
  updateStock: boolean;
}

export interface PurchaseBill {
  id: string;
  billNo: string;
  supplierId?: string;
  supplierName: string;
  supplierGSTIN?: string;
  supplierInvoiceNo?: string;
  linkedPOId?: string;
  status: BillStatus;
  items: PurchaseBillItem[];
  subtotal: number;
  totalCGST: number;
  totalSGST: number;
  totalIGST: number;
  totalGST: number;
  grandTotal: number;
  amountPaid: number;
  balanceDue: number;
  billDate: string;
  dueDate?: string;
  paymentMode?: PurchaseBillPaymentMode;
  paymentDate?: string;
  paymentStatus?: PurchaseBillPaymentStatus;
  paidVia?: PurchaseBillPaidVia;
  reminderEnabled?: boolean;
  remindOn?: string;
  reminderFreq?: PurchaseBillReminderFrequency;
  remindVia?: PurchaseBillReminderChannel;
  reminders?: Array<{ days: number; date: string; triggered: boolean }>;
  notes?: string;
  createdBy: string;
  createdAt: string;
}

export interface PayablesSummary {
  totalPayable: number;
  paidThisMonth: number;
  pendingBills: number;
  overdueBills: number;
}

File: pos-backend/src/main/java/com/retailflow/pos/purchasebill/PurchaseBill.java
package com.retailflow.pos.purchasebill;

import com.retailflow.pos.purchaseorder.PurchaseOrder;
import com.retailflow.pos.supplier.Supplier;
import jakarta.persistence.*;
import java.math.BigDecimal;
import java.time.*;
import java.util.*;
import lombok.*;

@Data @Builder @NoArgsConstructor @AllArgsConstructor
@Entity @Table(name="purchase_bills")
public class PurchaseBill {
  public enum Status { DRAFT, CONFIRMED, PAID, PARTIAL_PAID, CANCELLED }
  @Id private String id;
  @Column(nullable=false,unique=true) private String billNo;
  @ManyToOne(fetch=FetchType.LAZY) @JoinColumn(name="supplier_id") private Supplier supplier;
  private String supplierName;
  private String supplierGstin;
  private String supplierInvoiceNo;
  @ManyToOne(fetch=FetchType.LAZY) @JoinColumn(name="linked_po_id") private PurchaseOrder linkedPo;
  @Enumerated(EnumType.STRING) private Status status;
  private BigDecimal subtotal;
  private BigDecimal totalCgst;
  private BigDecimal totalSgst;
  private BigDecimal totalIgst;
  private BigDecimal totalGst;
  private BigDecimal grandTotal;
  private BigDecimal amountPaid;
  private BigDecimal balanceDue;
  private LocalDate billDate;
  private LocalDate dueDate;
  private String paymentMode;
  private LocalDate paymentDate;
  private String paymentStatus;
  private String paidVia;
  private LocalDateTime stockAppliedAt;
  private LocalDateTime stockReversedAt;
  private boolean reminderEnabled;
  private LocalDate remindOn;
  @Lob private String notes;
  private String createdBy;
  private LocalDateTime createdAt;
  @OneToMany(mappedBy="purchaseBill",cascade=CascadeType.ALL,orphanRemoval=true)
  @Builder.Default private List<PurchaseBillItem> items=new ArrayList<>();
  @OneToMany(mappedBy="purchaseBill",cascade=CascadeType.ALL,orphanRemoval=true)
  @Builder.Default private List<PurchaseBillAttachment> attachments=new ArrayList<>();
  @PrePersist void create(){if(id==null)id=UUID.randomUUID().toString();createdAt=LocalDateTime.now();if(status==null)status=Status.DRAFT;if(amountPaid==null)amountPaid=BigDecimal.ZERO;}
  public void setItems(List<PurchaseBillItem> values){items.clear();if(values!=null)values.forEach(i->{i.setPurchaseBill(this);items.add(i);});}
}
